diff --git a/src/ThemeSwitch.astro b/src/ThemeSwitch.astro
index 14e15743670e41747688d6faa9ce5dc41e0a5ec4..f9ecd4e3b2f583c05e2de9f39996a84d24f07bdd 100644
--- a/src/ThemeSwitch.astro
+++ b/src/ThemeSwitch.astro
@@ -19,7 +19,10 @@ const { strategy, defaultTheme, as: Element = "span" }: Props = Astro.props;
 </style>
 
 <script define:vars={{ strategy, defaultTheme }}>
-  const themeSwitch = document.getElementById("astro-color-scheme-switch");
+let document_ = document
+
+function task() {
+  const themeSwitch = document_.getElementById("astro-color-scheme-switch");
   const theme = localStorage.getItem("theme");
   const themeMatcher = window.matchMedia("(prefers-color-scheme: dark)");
   let systemTheme = themeMatcher.matches ? "dark" : "light";
@@ -33,9 +36,9 @@ const { strategy, defaultTheme, as: Element = "span" }: Props = Astro.props;
   });
 
   function updateAppliedTheme(value) {
-    document.documentElement.classList.remove("light", "dark");
-    document.documentElement.classList.add(value);
-    document.documentElement.setAttribute("data-theme", value);
+    document_.documentElement.classList.remove("light", "dark");
+    document_.documentElement.classList.add(value);
+    document_.documentElement.setAttribute("data-theme", value);
   }
 
   function updateTheme(value) {
@@ -106,4 +109,16 @@ const { strategy, defaultTheme, as: Element = "span" }: Props = Astro.props;
     default:
       updateTheme(theme || defaultTheme || systemTheme);
   }
+}
+
+task()
+
+document_.addEventListener('astro:before-swap', (event) => {
+  document_ = event.newDocument
+  task()
+
+  event.viewTransition.updateCallbackDone.then(() => {
+    document_ = document
+  });
+})
 </script>
